distribution/lib/Standard/Table/0.0.0-dev/src/IO/Auto_Detect.enso

```diff
@@ -0,0 +1,36 @@
+from Standard.Base import Any, Problem_Behavior, Nothing, Error, Panic, Meta, File, File_Format, Plain_Text, Bytes
+from Standard.Base.Error.Common import Unsupported_File_Type_Data, No_Such_Method_Error_Data, Illegal_Argument_Error_Data
+
+from Standard.Table.Delimited.Delimited_Format import Delimited
+from Standard.Table.Excel.Excel_Format import Excel
+
+types = [Plain_Text, Delimited ',', Excel, Bytes]
+
+type Auto_Detect
+    ## If the File_Format supports reading from the file, return a configured instance.
+    can_read : File -> Any
+    can_read self file =
+        reader idx =
+            formatter = types.at idx . can_read file
+            if formatter.is_nothing.not then formatter else
+                if idx + 1 == types.length then Nothing else
+                    @Tail_Call reader (idx + 1)
+        reader 0
```

Comment:
This feels a bit like we should have a relevant function in `Vector` instead of writing the loop here, although I'm not sure what this function should be called. We can always do `types.map .can_read . find (x -> x.is_nothing.not)` - but as our `map` is not lazy this will not terminate early.

If we want to keep this function, can we do
```suggestion
        reader idx =
            if idx >= types.length then Nothing else
                formatter = types.at idx . can_read file
                if formatter.is_nothing.not then formatter else
                    @Tail_Call reader (idx + 1)
        reader 0
```

I prefer the length check at the beginning in case the `types` were empty (I know they aren't but still...) and then the aritmetic is simpler (no idx+1) and its more conistent with other loops like that that we write - which I think is nice because you get used to such patterns and in time I recognize such patterns almost automatically. Whereas with the other ordering I looked at the loop and had to think a bit more to see what is going on.